mm.c

```diff
@@ -523,19 +523,74 @@ void mm_free(void *ptr)
  */
 void *mm_realloc(void *ptr, size_t size)
 {
-    void *oldptr = ptr;
-    void *newptr;
-    size_t copySize;
-
-    newptr = mm_malloc(size);
-    if (newptr == NULL)
-      return NULL;
-    copySize = *(size_t *)((char *)oldptr - 4);
-    if (size < copySize)
-      copySize = size;
-    memcpy(newptr, oldptr, copySize);
-    mm_free(oldptr);
-    return newptr;
+	// equivalent to mm_malloc if ptr is NULL
+	if (ptr == NULL) {
+		return mm_malloc(ptr);
+	}
+
+	// adjust to be at start of block
+	char **old = (char **)ptr - 1;
+	char **bp = (char **)ptr - 1;
+
+	// get intended and current size
+	size_t new_size = ALIGN(size) / WORD_SIZE; // in words
+	size_t old_size = GET_SIZE(bp); // in words
+
+	// equivalent to mm_free if size == 0
+	if (new_size == 0) {
+		mm_free(ptr);
+		return NULL;
+	} else if (new_size < old_size) { // checks if can avoid reallocating
+		PUT_WORD(bp, PACK(old_size, FREE));
+    PUT_WORD(FTRP(bp), PACK(old_size, FREE));
+		alloc_free_block(bp, new_size);
```

Comment:
wouldn't this lead to external fragmentation as bp+(old_size - new_size) to bp+old_size free space, but not in free block ?